### PR TITLE
UCP/GTEST: Support the wakeup mechanism with client-server and test it.

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

File: src/ucp/core/ucp_listener.c [hunk 1]
```diff
@@ -157,6 +161,11 @@ ucs_status_t ucp_listener_create(ucp_worker_h worker,
             goto err_free;
         }
 
+        if ((context->config.features & UCP_FEATURE_WAKEUP) &&
+            !(listener->wiface.attr.cap.flags & UCT_IFACE_FLAG_CB_ASYNC)) {
+            goto err_iface_cleanup;
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/core/ucp_listener.c]: so can we cleanup and then try next md?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6392/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6428/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6434/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe Can you please review?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 2]
```diff
@@ -66,7 +67,22 @@ class test_ucp_sockaddr : public ucp_test {
         ASSERT_UCS_OK(status);
     }
 
-    void tag_send_recv(entity& from, entity& to) {
+    void wait(ucp_worker_h worker, void *request, bool wakeup)
+    {
+        if (wakeup) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: can be written shorter, also isn't progressing all time time causing the test to "hang"?
```
do {
     if (wait) {
        ucp_worker_wait(worker);
     } else {
        sched_yield();
     }
     while (progress());
} while (!ucp_request_is_completed(request));
```

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 3]
```diff
@@ -109,13 +121,39 @@ class test_ucp_sockaddr : public ucp_test {
         ep_params.sockaddr.addrlen = sizeof(*connect_addr);
         sender().connect(&receiver(), ep_params);
 
-        tag_send_recv(sender(), receiver());
-
-        /* wait for reverse ep to appear */
-        while (receiver().get_num_eps() == 0) {
+        if (wakeup) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: there should not be any different path for "wait" in this function, only "wait()" function should be different.
why need to do wait/progress here if it's already done in wait()?

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 2 above)
Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: can you please clarify the question?

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 3 above)
Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: the wait function takes care of progressing after the send-recv starts. waiting here is for the connection establishment part on the server side.

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 2 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: you mentioned before that doing progress in a loop without sleep/yield causes tarvation of progress thread. but it is done here in case of wakeup==false

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 3 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: why need to wait for connection establishment?

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 2 above)
Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: the hang happened on the wakeup=true case. when the connection wasn't established yet when the sender was waiting for a completion on the send but didn't receive completion on the wireup message yet since the receiver probably didnt get the wireup connection request yet. 
this is why i wait for a connection establishment event first (answering the question from the other comment here as well)

Comment (yosefe) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: ok,  but now the sender is going to sleep by doing worker_wait(), we should be able to remove the code which waits for wireup

Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: the code that "waits" for wireup isnt in the 'wait' function. it's the short_progress_loop() in connect_and_send_recv() before even starting send-recv. 
if i remove it, the test may hang since send-recv will start before connection is established (since no one progressed the server side to make sure it reveived the wireup connection request).

Comment (yosefe) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: but according to UCP APi definition, the user is allowed to start sending immediately after client ep is created. only the receive should wait until server ep is created.

Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: ok, then inside connect_and_send_recv(), need this?
```
if (wakeup) {
            ucp_worker_wait(receiver().worker());
            while (progress());       --> while (receiver().progress()); 
            short_progress_loop();    --> remove

            tag_send_recv(sender(), receiver() ,true);
            tag_send_recv(receiver(), sender(), true);
        }
```

how can we make sure though that the server is connected (wireup wise)? and not only that it had the connection establishment event (client-server wise)?

Comment (yosefe) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: no need for the above. 
Need to progress only as part of waiting for send or receive request completion, or waiting for server ep to appear.
In this specific test, seems there is **no need** to make sure server is connected (or waiting for server ep to appear), because receive is done on the worker (tag api).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6507/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe please review

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -1457,26 +1457,32 @@ void ucp_worker_wait_mem(ucp_worker_h worker, void *address);
  *
  * @code {.c}
  * void application_initialization() {
+ * // should be called once in application init flow and before
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: can you please put the doc update in another PR? we want to review the api/doc change separately and not make the rdma_cm wakeup fix/test depend on the doc fix

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 5]
```diff
@@ -67,17 +68,49 @@ class test_ucp_sockaddr : public ucp_test {
         ASSERT_UCS_OK(status);
     }
 
-    void wait(ucp_worker_h worker, void *request, bool wakeup)
+    static void handle_wakeup(ucp_worker_h send_worker, ucp_worker_h recv_worker)
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: maybe rename this function to "wait_wakeup" or "wait_for_wakup"?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 6]
```diff
@@ -67,17 +68,49 @@ class test_ucp_sockaddr : public ucp_test {
...
+            ret = poll(pfd, 2, -1);
+        } while ((ret < 0) && (errno == EINTR));
+        if (ret < 0) {
+            UCS_TEST_MESSAGE << "poll() failed: " << strerror(errno);
+        }
+
+        EXPECT_GE(ret, 1);
+    }
+
+    void wait(ucp_worker_h send_worker, ucp_worker_h recv_worker, bool wakeup)
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: this does not really "wait" for something since it does a single progress and possible goes to sleep with wakeup,
maybe rename it to "progress_with_wakeup" or similar?

Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: but it calls a function which would be named "wait_for_wakeup".
also, since this current 'wait' is called not only when wakeup is used, do you think it's ok to use  'wakeup' in the name?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6520/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe can you please review again?

**Comment by yosefe** [comment]
@alinask pls squash
@shamisp @bbenton @gmegan pls check the small documentation update in uct.h

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6522/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6523/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6524/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@yosefe squashed.

File: src/uct/api/uct.h [hunk 7]
```diff
@@ -1148,8 +1148,8 @@ ucs_status_t uct_ep_check(const uct_ep_h ep, unsigned flags,
  * @ingroup UCT_RESOURCE
  * @brief Obtain a notification file descriptor for polling.
  *
- * Only interfaces supporting the @ref UCT_IFACE_FLAG_EVENT_FD implement this
- * function.
+ * Only interfaces supporting any of the UCT_IFACE_FLAG_EVENT* flags implement
```
Comment (gmegan) [inline src/uct/api/uct.h]: Different grammar would be more clear:
"Only interfaces that support at least one of the UCT_IFACE_FLAG_EVENT* flags will implement"

**Comment by gmegan** [comment]
See inline comment for wording change on documentation in uct.h. Otherwise all looks good.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6576/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@gmegan I applied you comment. Can you please take a look again?
